modules/internal/localeModels/gpu/LocaleModel.chpl

```diff
@@ -0,0 +1,345 @@
+
+      extern proc cudaGetDeviceCount(ref n: int);
+      var nDevices: int;
+      cudaGetDeviceCount(nDevices);
+
+      //1 cpu and number of GPU devices on a node
+      numSublocales = 1 + nDevices;
+      childSpace = {0..#numSublocales};
+
+      this.complete();
+
+      setup();
+    }
+
+    proc init(parent_loc : locale) {
+      if rootLocaleInitialized {
+        halt("Cannot create additional LocaleModel instances");
+      }
+
+      super.init(parent_loc);
+
+      _node_id = chpl_nodeID: int;
+
+      extern proc cudaGetDeviceCount(ref n: int);
+      var nDevices: int;
+      cudaGetDeviceCount(nDevices);
+
+      //1 cpu and number of GPU devices on a node
+      numSublocales = 1 + nDevices;
+      childSpace = {0..#numSublocales};
+
+      this.complete();
+
+      setup();
+    }
+
+    // top-level locale (node) number
+    override proc chpl_id() return _node_id;
+
+    override proc chpl_localeid() {
+      return chpl_buildLocaleID(_node_id:chpl_nodeID_t, c_sublocid_any);
+    }
+    override proc chpl_name() return local_name;
+
+    proc getChildSpace() return childSpace;
+
+    override proc getChildCount() return numSublocales;
+
+    iter getChildIndices() : int {
+      for idx in {0..#numSublocales} do // chpl_emptyLocaleSpace do
+        yield idx;
+    }
+
+    override proc getChild(idx:int) : locale {
+      if boundsChecking then
+        if (idx < 0) || (idx >= numSublocales) then
+          halt("sublocale child index out of bounds (",idx,")");
+      return new locale(childLocales[idx]);
+    }
+
+    iter getChildren() : locale  {
+      for loc in childLocales do
+        yield new locale(loc);
+    }
+
+    proc getChildArray() {
+      return [loc in childLocales] loc;
+    }
+
+    //------------------------------------------------------------------------{
+    //- Implementation (private)
+    //-
+    proc setup() {
+      helpSetupLocaleGPU(this, numSublocales, local_name, CPULocale, GPULocale);
+    }
+    //------------------------------------------------------------------------}
+  }
+
+  //
+  // An instance of this class is the default contents 'rootLocale'.
+  //
+  // In the current implementation a platform-specific architectural description
+  // may overwrite this instance or any of its children to establish a more customized
+  // representation of the system resources.
+  //
+  class RootLocale : AbstractRootLocale {
+
+    const myLocaleSpace: domain(1) = {0..numLocales-1};
+    pragma "unsafe"
+    var myLocales: [myLocaleSpace] locale;
+
+    proc init() {
+      super.init(nilLocale);
+      nPUsPhysAcc = 0;
+      nPUsPhysAll = 0;
+      nPUsLogAcc = 0;
+      nPUsLogAll = 0;
+      maxTaskPar = 0;
+    }
+
+    // The setup() function must use chpl_initOnLocales() to iterate (in
+    // parallel) over the locales to set up the LocaleModel object.
+    // In addition, the initial 'here' must be set.
+    proc setup() {
+      helpSetupRootLocaleNUMA(this);
```

Comment:
This might answer my earlier question about `helpSetupRootLocaleGPU()`, above.  My guess is that it won't work to be calling the NUMA helper here, because the sublocale architecture is different.  So we probably really do need to implement the GPU helper and call it here?

Reply:
The `helpSetupRootLocaleNuma` function (https://github.com/chapel-lang/chapel/blob/master/modules/internal/LocaleModelHelpSetup.chpl#L80) and `helpSetupRootLocaleAPU` function (https://github.com/chapel-lang/chapel/blob/master/modules/internal/LocaleModelHelpSetup.chpl#L93) look the same except the APU function has
```here.runningTaskCntSet(0);  // locale init parallelism mis-sets this``` at the end. Do you remember what `here.runningTaskCntSet(0);` does? I'm trying to figure out if it'll be necessary in the new `helpSetupRootLocaleGPU()` function.

Reply:
Check with @ronawho to be certain, but my guess is that whatever reason we needed to set it in the APU locale model may hold for the GPU one also.  Either that, or it's not needed in both cases!  (That's the root of my uncertainty.)

Reply:
I think it's not needed in both cases. 

`here.runningTaskCntSet` sets the "running task counter" for the current node. That's just a count of how many tasks are currently running and we use it for things like throttling nested parallelism. e.g. if you have:

```chpl
forall 1..1000 do
  forall 1..1000 do 
    func();
```
We don't create tasks for the inner forall since we're already fully parallel. 

There shouldn't be any reason to reset it to 0 here (we do that elsewhere)

Reply:
Got it, thanks. I'll leave the`helpSetupRootLocaleGPU()` function as is then (which currently does not have `here.runningTaskCntSet(0)` at the end).